src/GlobalState.lua

```diff
@@ -58,6 +70,69 @@ function M.inner_setup()
   end
 end
 
+-- mark an item as missing
+function M.missing_item_set(item)
+  global.mod.missing_item[item] = game.tick
+end
+
+-- return true if an item has been marked as missing
+function M.missing_item_get(item)
+  return global.mod.missing_item[item] ~= nil
+end
+
+local function missing_filter(tab)
+  local max_delta = constants.MAX_MISSING_TICKS
+  local to_del = {}
+  local now = game.tick
+  for name, tick in pairs(tab) do
+    if (now - tick) > max_delta then
+      table.insert(to_del, name)
+    end
+  end
+  for _, name in ipairs(to_del) do
+    tab[name] = nil
+  end
+  return tab
+end
+
+-- drop any items that have not been missing for a while
+-- returns the (read-only) table of missing items
+function M.missing_item_filter()
+  return missing_filter(global.mod.missing_item)
+end
+
+-- create a string 'key' for a fluid@temp
+function M.fluid_temp_key(fluid_name, temp)
+  return string.format("%s;%d", fluid_name, math.floor(temp * 1000))
```

Comment:
What do you think about this fluid string encoding vs the two-table approach used to track fluid counts? Do you think there will be any issues when rounding a fluid's temp?

I don't think there will be issues, and the stakes for an incorrect fluid temp are lower than the actual fluid counts, but it would be nice long-term to use the same strategy for storing temperatures throughout the mod.

Reply:
I combined the fluid and temp for three reasons:

1) I've had trouble using a floating point number as a key in a Lua table in other projects. (25 vs 25.0001 vs 25.0002). I always ended up encoding it as a string with fixed precision.
So, if I needed 0.1 precision, I'd use math.floor(val*10) to encode the key. Sometimes as a hex string to get a constant width.

2) the code to build the two-layered tables got a little verbose. Especially on cleanup.

3) I can re-use the same 'timeout' code for items and fluids, as the table structure is the same


As a side note, there are really only a few "useful" temperatures for fluids. (regular steam, hot steam and everything else). We'd need to do something with the temperature to normalize it.
Creating a new item name seems like an easy approach to hide the complexity.
"steam@500" vs "steam@165".
And if the temperature of the fluid doesn't matter, then we could leave it off.

Reply:
Good to know that you ran into issues with floating point numbers as a key in other projects. I'll keep an eye out in case that pops up here.

I haven't heard of non-integer temperatures being using in factorio so rounding makes sense.

Reply:
Is there anything other than "steam" that benefits from the temperature?
If we are only doing integers, then there is no need to multiply by 1000.
So far, 165 and 500 seem to be useful.